### PR TITLE
Prevent stream executor construction for XLA devices

### PR DESCRIPTION
**Pull request by samikama**
This PR prevents stream executor construction for the devices that are not in the visible_gpu_devices list given by the user. Construction of executors was causing allocations on other devices, reducing available memory.

### CODE REVIEW

**Comment by jlebar** [comment]
At a high level, I don't think `std::set<int>` where `{-1}` means "all devices" is the cleanest abstraction for representing this constraint.

I think preferable would be `absl::optional<std::set<Int>>` in the `Options` struct, and then as soon as we can enumerate all the device ordinals, we convert this into a plain `std::set<int>` without the special case for `{-1}` and just fill the set with all devices.

WDYT?

**Review by jlebar — CHANGES_REQUESTED** [review]

**Comment by samikama** [comment]
I didn't thought of optional. I will convert it and address other comment and update tomorrow. Thanks.

File: tensorflow/compiler/jit/xla_device.h [hunk 1]
```diff
@@ -256,6 +259,9 @@ class XlaDevice : public LocalDevice {
   // completion.
   int64 outstanding_asynchronous_operations_ GUARDED_BY(mu_) = 0;
   condition_variable outstanding_asynchronous_operations_cv_;
+
+  // Set of allowed gpu devices at the time of construction.
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: Yes. There is no way that I am aware of to control the CPU devices in TF.

[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: A different session configuration may be passed to another session. It should not change throughout the session lifetime.

[the remark this answers is not in the document]
Comment (jlebar) [reply tensorflow/compiler/jit/xla_device.h]: XLA works for devices other than GPU and CPU.

Can you rephrase the comment?

[the remark this answers is not in the document]
Comment (jlebar) [reply tensorflow/compiler/jit/xla_device.h]: If the comment is misleading, would you be willing to revise it?

File: tensorflow/compiler/xla/service/platform_util.cc [hunk 2]
```diff
@@ -226,6 +227,11 @@ PlatformUtil::GetStreamExecutors(se::Platform* platform) {
     tensorflow::thread::ThreadPool thread_pool(
         tensorflow::Env::Default(), "device_initialization", device_count);
     for (int i = 0; i < device_count; ++i) {
+      if (allowed_devices.count(-1) == 0 && allowed_devices.count(i) == 0) {
+        VLOG(1) << "Skipping stream executor for device " << i
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/service/platform_util.cc]: Done

File: tensorflow/compiler/jit/xla_device.h [hunk 3]
```diff
@@ -123,6 +124,8 @@ class XlaDevice : public LocalDevice {
     // If padded_shape_fn is empty, a default implementation that returns
     // the logical on-device shape without padding is used.
     PaddedShapeFn padded_shape_fn;
+    // Set of allowed devices. -1 is all devices
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: Done

File: tensorflow/compiler/jit/xla_device.h [hunk 4]
```diff
@@ -123,6 +124,8 @@ class XlaDevice : public LocalDevice {
...
+    std::set<int> allowed_devices = {-1};
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: I tried to improve the comment. Let me know if you would like something different.

File: tensorflow/compiler/jit/xla_device.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: I tried to address this. Please let me know any improvements you would like to see.

[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: done

File: tensorflow/compiler/xla/client/client_library.cc [hunk 5]
```diff
@@ -58,6 +60,16 @@ int LocalClientOptions::intra_op_parallelism_threads() const {
   return intra_op_parallelism_threads_;
 }
 
+LocalClientOptions& LocalClientOptions::set_allowed_devices(
+    std::set<int> device_set) {
+  allowed_devices_ = device_set;
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/client/client_library.cc]: Changed function signature.

File: tensorflow/compiler/xla/client/client_library.cc [hunk 6]
```diff
@@ -58,6 +60,16 @@ int LocalClientOptions::intra_op_parallelism_threads() const {
...
+  return *this;
+}
+
+std::set<int> LocalClientOptions::get_allowed_devices() const {
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/client/client_library.cc]: done

File: tensorflow/compiler/xla/client/client_library.cc [hunk 7]
```diff
@@ -67,9 +79,10 @@ ClientLibrary::ClientLibrary() = default;
 ClientLibrary::~ClientLibrary() = default;
 
 /* static */ StatusOr<LocalClient*> ClientLibrary::GetOrCreateLocalClient(
-    se::Platform* platform) {
+    se::Platform* platform, const std::set<int> device_set) {
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/client/client_library.cc]: Changed these to take const reference.

File: tensorflow/compiler/xla/client/client_library.h [hunk 8]
```diff
@@ -45,7 +46,8 @@ class LocalClientOptions {
  public:
   LocalClientOptions(se::Platform* platform = nullptr,
                      int number_of_replicas = 1,
-                     int intra_op_parallelism_threads = -1);
+                     int intra_op_parallelism_threads = -1,
+                     std::set<int> device_set = {-1});
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/client/client_library.h]: Done.

File: tensorflow/compiler/xla/client/client_library.h [hunk 9]
```diff
@@ -60,10 +62,16 @@ class LocalClientOptions {
   LocalClientOptions& set_intra_op_parallelism_threads(int num_threads);
   int intra_op_parallelism_threads() const;
 
+  // Sets the allowed_devices set for creation of stream executors.
+  LocalClientOptions& set_allowed_devices(const std::set<int> device_set);
+
+  std::set<int> get_allowed_devices() const;
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/client/client_library.h]: Done.

File: tensorflow/compiler/xla/service/backend.h [hunk 10]
```diff
@@ -106,6 +113,12 @@ class Backend {
   // can be > 1).
   se::StreamExecutor* default_stream_executor() const {
     CHECK(!stream_executors_.empty());
+
+    for (se::StreamExecutor* e : stream_executors_) {
+      if (e) {
+        return e;
+      }
+    }
     return stream_executors_[0];
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/service/backend.h]: I changed this method. It was a forgotten artifact from when I was trying a different way to filter the device set. It is not possible to have nullptr in the vector now.

File: tensorflow/compiler/xla/service/backend.h [hunk 11]
```diff
@@ -106,6 +113,12 @@ class Backend {
   // can be > 1).
   se::StreamExecutor* default_stream_executor() const {
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/compiler/xla/service/backend.h]: They are non-null again.

**Comment by samikama** [comment]
@jlebar, please let me know if you have any further comments and I will try to get them in before the branching tomorrow.

Thanks,
Sami

File: tensorflow/compiler/jit/xla_device.h [hunk 12]
```diff
@@ -124,8 +125,11 @@ class XlaDevice : public LocalDevice {
     // If padded_shape_fn is empty, a default implementation that returns
     // the logical on-device shape without padding is used.
     PaddedShapeFn padded_shape_fn;
-    // Set of allowed devices. -1 is all devices
-    std::set<int> allowed_devices = {-1};
+
+    // Set of devices to use. This controls which of the devices given type in
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_device.h]: Typo `devices given type`?

File: tensorflow/compiler/jit/xla_device.h [hunk 13]
```diff
@@ -124,8 +125,11 @@ class XlaDevice : public LocalDevice {
...
+    // the system will have resources allocated for. For GPUs this will be
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_device.h]: typo `in the system will have resources allocated for`?

File: tensorflow/compiler/xla/client/client_library.h [hunk 14]
```diff
@@ -47,7 +48,7 @@ class LocalClientOptions {
   LocalClientOptions(se::Platform* platform = nullptr,
                      int number_of_replicas = 1,
                      int intra_op_parallelism_threads = -1,
-                     std::set<int> device_set = {-1});
+                     const absl::optional<std::set<int>>& allowed_devices = {});
```
Comment (jlebar) [inline tensorflow/compiler/xla/client/client_library.h]: Nit, prefer `nullopt` ("explicit is better than implicit", and `{}` looks like we're initializing the set to an empty set, which is the opposite of what we're doing.)

File: tensorflow/compiler/xla/service/platform_util.cc [hunk 15]
```diff
@@ -227,8 +228,8 @@ PlatformUtil::GetStreamExecutors(se::Platform* platform,
     tensorflow::thread::ThreadPool thread_pool(
         tensorflow::Env::Default(), "device_initialization", device_count);
     for (int i = 0; i < device_count; ++i) {
-      if (allowed_devices.count(-1) == 0 && allowed_devices.count(i) == 0) {
-        VLOG(1) << "Skipping stream executor for device " << i
+      if (allowed_devices && (*allowed_devices).count(i) == 0) {
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/platform_util.cc]: `allowed_devices->count`?

File: tensorflow/compiler/xla/service/platform_util.h [hunk 16]
```diff
@@ -64,7 +64,8 @@ class PlatformUtil {
   //
   // If the platform has no visible devices, a not-found error is returned.
   static StatusOr<std::vector<se::StreamExecutor*>> GetStreamExecutors(
-      se::Platform* platform, std::set<int> allowed_devices = {-1});
+      se::Platform* platform,
+      const absl::optional<std::set<int>>& allowed_devices = {});
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/platform_util.h]: `= nullopt`

Comment (jlebar) [inline tensorflow/compiler/xla/service/platform_util.h]: Please comment what the `allowed_devices` argument does and why it's there.  (That is, why can't we just take the resulting vector and *then* filter it?  That's the whole point of this patch, actually; it's quite subtle, and right now it's not explained anywhere in the code.)

File: tensorflow/compiler/xla/service/service.h [hunk 17]
```diff
@@ -63,15 +63,16 @@ class ServiceOptions {
   int intra_op_parallelism_threads() const;
 
   // Sets the allowed_devices set for creation of stream executors.
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/service.h]: Same comment as in last review about comments that repeat verbatim what's in the code.  This occurs in a few places in this patch.

I'm not totally wild about repeating the comment explaining what allowed_devices means everywhere, but since it's only 3 lines or so, maybe that's better than having a pointer in the comments (i.e. it's better than "see Foo::set_allowed_devices()").

**Review by jlebar — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/platform_util.h (hunk 16 above)
Comment (samikama) [reply tensorflow/compiler/xla/service/platform_util.h]: Done.

File: tensorflow/compiler/xla/service/service.h (hunk 17 above)
Comment (samikama) [reply tensorflow/compiler/xla/service/service.h]: `Same comment as in last review about comments that repeat verbatim what's in the code. This occurs in a few places in this patch.`, assuming you are talking about the space between getters-setters, done. If not please clarify what do you want me to do.

File: tensorflow/compiler/jit/xla_device.h [hunk 18]
```diff
@@ -126,8 +126,8 @@ class XlaDevice : public LocalDevice {
     // the logical on-device shape without padding is used.
     PaddedShapeFn padded_shape_fn;
 
-    // Set of devices to use. This controls which of the devices given type in
-    // the system will have resources allocated for. For GPUs this will be
+    // Set of devices to use. This controls which of the devices on the given
+    // platform resources will have resources allocated. For GPUs this will be
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_device.h]: > This controls which of the devices on the given
> platform resources will have resources allocated.

Should the first "resources" be removed?

File: tensorflow/compiler/xla/service/platform_util.cc [hunk 19]
```diff
@@ -228,7 +228,13 @@ PlatformUtil::GetStreamExecutors(
     tensorflow::thread::ThreadPool thread_pool(
         tensorflow::Env::Default(), "device_initialization", device_count);
     for (int i = 0; i < device_count; ++i) {
-      if (allowed_devices && (*allowed_devices).count(i) == 0) {
+      // Once a stream executor is instantiated it will cause allocations on
+      // the device, for example for GPUs cuda context, cudnn handles etc. will
+      // be constructed. By constructing stream executors only on the
+      // allowed_devices, we don't make any allocations on other devices.
+      // This helps in multi-process executions on the same host like horovod or
+      // shared hosts.
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/platform_util.cc]: This is a great comment, thank you.

File: tensorflow/compiler/xla/service/platform_util.h [hunk 20]
```diff
@@ -61,11 +61,13 @@ class PlatformUtil {
   // Returns a vector of StreamExecutors for the given platform. The vector is
   // indexed by device ordinal (device numbering used by StreamExecutor). If an
   // element is nullptr, then the device is present by not supported by XLA.
+  // Optional parameter, allowed_devices controls which of the devices on the
+  // platform will have StreamExecutors constructed for.
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/platform_util.h]: No need to say it's an optional parameter, that is clear from the code.

>  allowed_devices controls which of the devices on the platform will have StreamExecutors constructed for. 

Reword sentence?

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 21]
```diff
@@ -33,8 +33,34 @@ class XlaGpuDeviceFactory : public DeviceFactory {
  public:
   Status CreateDevices(const SessionOptions& options, const string& name_prefix,
                        std::vector<std::unique_ptr<Device>>* devices) override;
+  // Returns a set containing the device ids contained in visible_device_list or
+  // -1 if the string is empty.
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: Should we be returning a `StatusOr<optional<set<int>>>` instead of returning {-1}?

But also I'm confused, it looks like we never call this function when the input string is empty?  So it never actually returns nullopt or {-1}.  Seems like we could simplify its contract?

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 22]
```diff
@@ -33,8 +33,34 @@ class XlaGpuDeviceFactory : public DeviceFactory {
...
+  static xla::StatusOr<std::set<int>> ParseVisibleDeviceList(
+      const string& visible_device_list);
 };
 
+xla::StatusOr<std::set<int>> XlaGpuDeviceFactory::ParseVisibleDeviceList(
+    const string& visible_device_list) {
+  std::set<int> gpu_ids;
+  if (visible_device_list.length() == 0) {
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: Sorry, now I am really confused.

Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: check `.empty()` instead of `length() == 0`

File: tensorflow/compiler/xla/client/client_library.h [hunk 23]
```diff
@@ -73,8 +82,11 @@ class ClientLibrary {
   //
   //   platform : The platform the underlying XLA service should target. If
   //     null then default platform is used.
+  //   device_set: Set of device IDs for which the stream executor will be
+  //   created for, for the given platform.
```
Comment (jlebar) [inline tensorflow/compiler/xla/client/client_library.h]: "for which X will be created for" -> "for which X will be created" or "which X will be created for"

File: tensorflow/compiler/xla/service/backend.cc [hunk 24]
```diff
@@ -172,7 +183,7 @@ StatusOr<se::StreamExecutor*> Backend::stream_executor(
         device_ordinal, stream_executors_.back()->device_ordinal());
   }
   for (auto* executor : stream_executors_) {
-    if (executor->device_ordinal() == device_ordinal) {
+    if (executor && executor->device_ordinal() == device_ordinal) {
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/backend.cc]: This null check can be reverted now, right?

**Review by jlebar — CHANGES_REQUESTED** [review]
Since we're having a good amount of back-and-forth on the comments, you may want to have someone who sits near you read over them before sending out the next round.

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 25]
```diff
@@ -34,17 +34,17 @@ class XlaGpuDeviceFactory : public DeviceFactory {
...
-xla::StatusOr<std::set<int>> XlaGpuDeviceFactory::ParseVisibleDeviceList(
-    const string& visible_device_list) {
+xla::StatusOr<absl::optional<std::set<int>>>
+XlaGpuDeviceFactory::ParseVisibleDeviceList(const string& visible_device_list) {
   std::set<int> gpu_ids;
-  if (visible_device_list.length() == 0) {
-    gpu_ids.insert(-1);
-    return gpu_ids;
+  if (visible_device_list.empty()) {
+    return absl::optional<std::set<int>>(absl::nullopt);
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: I think `return {nullopt}` or `return {{nullopt}}` may work.

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 26]
```diff
@@ -53,12 +53,12 @@ xla::StatusOr<std::set<int>> XlaGpuDeviceFactory::ParseVisibleDeviceList(
...
           "Could not parse entry in 'visible_device_list': '",
-          platform_gpu_id_str, "'. visible_device_list = ",
-          visible_device_list);
+          platform_gpu_id_str,
+          "'. visible_device_list = ", visible_device_list);
     }
     gpu_ids.insert(platform_gpu_id);
   }
-  return gpu_ids;
+  return absl::optional<std::set<int>>(gpu_ids);
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: Similarly, I think you should be able to `return {gpu_ids}` or maybe `return {{gpu_ids}}`.

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 27]
```diff
@@ -83,14 +83,16 @@ Status XlaGpuDeviceFactory::CreateDevices(
   }
   string allowed_gpus =
       session_options.config.gpu_options().visible_device_list();
+  auto parsed_gpus=ParseVisibleDeviceList(allowed_gpus).ValueOrDie();
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: Please format properly (or use clang-format)

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 28]
```diff
@@ -83,14 +83,16 @@ Status XlaGpuDeviceFactory::CreateDevices(
...
+  // We want to fill the gpu_ids set with all devices if config string is empty.
   std::set<int> gpu_ids;
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: If we're going to have ParseVisibleDeviceList return an optional, could we have just one variable?

```
absl::optional<std::set<int>> gpu_ids = ParseVisibleDeviceList(allowed_gpus);
if (!gpu_ids) {
  gpu_ids.emplace();
  for ... { gpu_ids->insert(i); }
}
for (int i : *gpu_ids) { ... }
```

File: tensorflow/compiler/xla/client/client_library.h [hunk 29]
```diff
@@ -60,10 +63,16 @@ class LocalClientOptions {
   LocalClientOptions& set_intra_op_parallelism_threads(int num_threads);
   int intra_op_parallelism_threads() const;
 
+  // Sets the allowed_devices set for creation of stream executors.
```
Comment (jlebar) [inline tensorflow/compiler/xla/client/client_library.h]: Change this one too per previous review?

File: tensorflow/compiler/xla/service/backend.h [hunk 30]
```diff
@@ -53,9 +54,15 @@ class BackendOptions {
   BackendOptions& set_intra_op_parallelism_threads(int num_threads);
   int intra_op_parallelism_threads() const;
 
+  // Sets the allowed_devices set for creation of stream executors.
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/backend.h]: Change this one too per previous review?

**Review by jlebar — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/service.h (hunk 17 above)
Comment (samikama) [reply tensorflow/compiler/xla/service/service.h]: Ok, I was confused by the comments above. I tried to improve it. I hope it is better, if not please suggest a better one and I will try to learn from that.

File: tensorflow/compiler/xla/service/platform_util.h (hunk 20 above)
Comment (samikama) [reply tensorflow/compiler/xla/service/platform_util.h]: Done.

File: tensorflow/compiler/xla/service/backend.cc (hunk 24 above)
Comment (samikama) [reply tensorflow/compiler/xla/service/backend.cc]: Done.

File: tensorflow/compiler/xla/client/client_library.h (hunk 23 above)
Comment (samikama) [reply tensorflow/compiler/xla/client/client_library.h]: Done.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 22 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Done.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 21 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Changed it to return optional.

File: tensorflow/compiler/jit/xla_device.h (hunk 18 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_device.h]: Done.

**Comment by samikama** [comment]
@jlebar I tried to address your comments. Please bear with me until I get used to the XLA style. Let me know if more changes are needed.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 28 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: I thought of this as well. But it is not much different that current implementation since if the string is empty, function will return a nullopt so we have to instantiate new set, put it in optional and fill it with all devices and then iterate over it. I can also do that but this way seemed cleaner.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 25 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: That was what I had initially but gcc couldn't be able to instantiate StatusOr from just nullopt.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 28 above)
Comment (jlebar) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: > so we have to instantiate new set, put it in optional and fill it with all devices and then iterate over it. I can also do that but this way seemed cleaner.

That's what the code I suggested above does, right?

Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Done.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 27 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Done. Though clang-format-5.0 formats this file differently that the clang-format in CI. I had to manually format some places.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 26 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Done

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 31]
```diff
@@ -33,8 +33,34 @@ class XlaGpuDeviceFactory : public DeviceFactory {
  public:
   Status CreateDevices(const SessionOptions& options, const string& name_prefix,
                        std::vector<std::unique_ptr<Device>>* devices) override;
+  // Returns a set containing the device ids contained in visible_device_list or
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: Can this be a standalone function rather than a class member?

Sorry I didn't see this earlier.

If you move it to a standalone function, please mark it as `static`.

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 32]
```diff
@@ -57,40 +83,24 @@ Status XlaGpuDeviceFactory::CreateDevices(
   }
   string allowed_gpus =
       session_options.config.gpu_options().visible_device_list();
-  std::set<int> gpu_ids;
+  absl::optional<std::set<int>> gpu_ids = ParseVisibleDeviceList(allowed_gpus).ValueOrDie();
+  // We want to fill the gpu_ids set with all devices if config string is empty.
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: s/We want to//

File: tensorflow/compiler/jit/xla_gpu_device.cc [hunk 33]
```diff
@@ -57,40 +83,24 @@ Status XlaGpuDeviceFactory::CreateDevices(
...
   int num_visible_devices = platform.ValueOrDie()->VisibleDeviceCount();
```
Comment (jlebar) [inline tensorflow/compiler/jit/xla_gpu_device.cc]: Move this line into the `if` statement?  Or even just change the loop to

```
for (int i = 0; i < platform.ValueOrDie()->VisibleDeviceCount(); ++i)
```

This has the advantage that now the comment on the line above is right above the if statement.  Makes a lot more sense.

Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: I forgot to update the comment when I made the earlier change. Sorry about that. It is fixed now.

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 32 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Done

File: tensorflow/compiler/jit/xla_gpu_device.cc (hunk 31 above)
Comment (samikama) [reply tensorflow/compiler/jit/xla_gpu_device.cc]: Done

**Review by jlebar — APPROVED** [review]
Approved, \o/

FYI re clang-format, it is not a stable format.  Newer versions of clang-format will more closely match what the CI expects.  But I believe our tooling will now clang-format patches when we import them, so I'm hoping we can disable the clang-format CI check altogether.  I just sent mail about it.

**Comment by jlebar** [comment]
Will get this merged asap so @tfboyd can consider a cherry-pick.

**Comment by DEKHTIARJonathan** [comment]
Seems like it solves:
- https://github.com/uber/horovod/issues/602
- https://github.com/tensorflow/tensorflow/issues/23458

 I will do a final check once merged and close the issues if I can confirm.

**Comment by samikama** [comment]
I believe it is a bug/feature in Grappler. At
https://github.com/tensorflow/tensorflow/blob/6dbe2cd22fa8d6b48ef9bb913c00de74886be07f/tensorflow/core/grappler/grappler_item_builder.cc#L106
grappler_item_builder is initializing a new set of devices from the device factory without passing session config.
According to https://github.com/tensorflow/tensorflow/issues/18861 initializing devices with different session config is an error condition. I am trying to understand the expected behavior. Perhaps we should make initialization sticky.
